src/udev/iocost/iocost.c

```diff
@@ -0,0 +1,314 @@
+                        items,
+                        CONFIG_PARSE_WARN,
+                        NULL,
+                        NULL);
+}
+
+static int help(void) {
+        printf("%s [OPTIONS...]\n\n"
+               "Set up iocost model and qos solutions for block devices\n"
+               "\nCommands:\n"
+               "  apply <path> [solution]    Apply the known solution for the device, if any, otherwise do nothing\n"
+               "  query <path>               Query the known solution for the device\n"
+               "\nOptions:\n"
+               "  -h --help                  Show this help\n"
+               "     --version               Show package version\n",
+               program_invocation_short_name);
+
+        return 0;
+}
+
+static int parse_argv(int argc, char *argv[]) {
+        enum {
+                ARG_VERSION = 0x100,
+        };
+
+        static const struct option options[] = {
+                { "help",      no_argument,       NULL, 'h'           },
+                { "version",   no_argument,       NULL, ARG_VERSION   },
+                {}
+        };
+
+        int c;
+
+        assert(argc >= 1);
+        assert(argv);
+
+        while ((c = getopt_long(argc, argv, "h", options, NULL)) >= 0)
+                switch (c) {
+
+                case 'h':
+                        return help();
+
+                case ARG_VERSION:
+                        return version();
+
+                case '?':
+                        return -EINVAL;
+
+                default:
+                        assert_not_reached();
+                }
+
+        return 1;
+}
+
+static int get_known_solutions(sd_device *device, char ***ret_solutions) {
+        _cleanup_free_ char **s = NULL;
+        const char *value;
+        int r;
+
+        assert(ret_solutions);
+
+        r = sd_device_get_property_value(device, "IOCOST_SOLUTIONS", &value);
+        if (r < 0)
+                return r;
+
+        s = strv_split(value, " ");
+        if (!s)
+                return -ENOMEM;
+
+        *ret_solutions = TAKE_PTR(s);
+
+        return 0;
+}
+
+static int choose_solution(char **solutions, const char **ret_name) {
+        assert(ret_name);
+
+        if (strv_isempty(solutions))
+                return log_error_errno(
+                                SYNTHETIC_ERRNO(EINVAL), "IOCOST_SOLUTIONS exists in hwdb but is empty.");
+
+        if (arg_bandwidth_threshold && strv_find(solutions, arg_bandwidth_threshold)) {
+                *ret_name = arg_bandwidth_threshold;
+                log_debug("Selected solution based on bandwidth threshold: %s", *ret_name);
+        } else {
+                *ret_name = solutions[0];
+                log_debug("Selected first available solution: %s", *ret_name);
+        }
+
+        return 0;
+}
+
+static int query_named_solution(
+                sd_device *device,
+                const char *name,
+                const char **ret_model,
+                const char **ret_qos) {
```

Comment:
another style nit: when breaking args in multiple lines, add an empty line after the `{` and before the first declaration